Online schema change best practices (#16942)
cockroachdb/docs: v23.1/online-schema-changes.md

If a schema change fails, the schema change job will be cleaned up automatically. However, there are limitations with rolling back schema changes within a transaction; for more information, see [Schema change DDL statements inside a multi-statement transaction can fail while other statements succeed](#schema-change-ddl-statements-inside-a-multi-statement-transaction-can-fail-while-other-statements-succeed).
{{site.data.alerts.end}}

> See [Estimate your storage capacity before performing online schema changes](#estimate-your-storage-capacity-before-performing-online-schema-changes) for information on how to avoid running out of space during an online schema change.

Reviewer: ```suggestion
For advice about how to avoid running out of space during an online schema change, refer to [Estimate your storage capacity before performing online schema changes](#estimate-your-storage-capacity-before-performing-online-schema-changes).
```

{% include_cached copy-clipboard.html %}
~~~ shell
SHOW RANGES FROM TABLE {table name} WITH DETAILS, KEYS, INDEXES;
~~~

The output includes a `range_size_mb` column that shows the size of the range in megabytes for each index.

In many cases this range size is trivial, but when the range size is many gigabytes or terabytes, you will need at least three times that amount of free storage space to successfully apply an online schema change.

#### Example of finding the range size of an index

1. Start a 3 node [`cockroach demo`](cockroach-demo.html) cluster with the MovR dataset.

    {% include_cached copy-clipboard.html %}
    ~~~ shell
    cockroach demo --nodes 3
    ~~~

1. Turn off the deprecated behavior of `SHOW RANGES`:

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    SET CLUSTER SETTING sql.show_ranges_deprecated_behavior.enabled TO 'false';
    ~~~

1. Find the range size of the indexes in the `movr.vehicles` table:

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    WITH x AS (
            SHOW RANGES FROM TABLE movr.vehicles WITH DETAILS, KEYS, INDEXES
         )
    SELECT index_name,

>            raw_index_start_key,

Reviewer 2: Maybe make this more user-friendly and let's not select keys

Author: Done.

## Best practices for online schema changes

> ### Estimate your storage capacity before performing online schema changes

Reviewer: If you take suggestions above, keep this in sync.
